src/deluge/gui/views/midi_session_view.cpp

```diff
@@ -0,0 +1,949 @@
+
+//initialize variables
+MidiSessionView::MidiSessionView() {
+	successfullyReadDefaultsFromFile = false;
+
+	anyChangesToSave = false;
+	onParamDisplay = false;
+	showLearnedParams = false;
+
+	initPadPress(lastPadPress);
+	initMapping(paramToCC);
+	initMapping(backupXMLParamToCC);
+	initMapping(previousKnobPos);
+	initMapping(previousCCValueSent);
+}
+
+void MidiSessionView::initPadPress(MidiPadPress& padPress) {
+	padPress.isActive = false;
+	padPress.xDisplay = kNoSelection;
+	padPress.yDisplay = kNoSelection;
+	padPress.paramKind = Param::Kind::NONE;
+	padPress.paramID = kNoSelection;
+}
+
+void MidiSessionView::initMapping(int32_t mapping[kDisplayWidth][kDisplayHeight]) {
+	for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+		for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
+			mapping[xDisplay][yDisplay] = kNoSelection;
+		}
+	}
+}
+
+bool MidiSessionView::opened() {
+	focusRegained();
+
+	return true;
+}
+
+void MidiSessionView::focusRegained() {
+	currentSong->affectEntire = true;
+
+	ClipNavigationTimelineView::focusRegained();
+	view.focusRegained();
+	view.setActiveModControllableTimelineCounter(currentSong);
+
+	if (!successfullyReadDefaultsFromFile) {
+		readDefaultsFromFile();
+	}
+
+	setLedStates();
+
+	updateMappingChangeStatus();
+
+	if (display->have7SEG()) {
+		redrawNumericDisplay();
+	}
+
+	uiNeedsRendering(this);
+}
+
+void MidiSessionView::graphicsRoutine() {
+	uint8_t tickSquares[kDisplayHeight];
+	uint8_t colours[kDisplayHeight];
+
+	// Nothing to do here but clear since we don't render playhead
+	memset(&tickSquares, 255, sizeof(tickSquares));
+	memset(&colours, 255, sizeof(colours));
+	PadLEDs::setTickSquares(tickSquares, colours);
+}
+
+ActionResult MidiSessionView::timerCallback() {
+	return ActionResult::DEALT_WITH;
+}
+
+bool MidiSessionView::renderMainPads(uint32_t whichRows, uint8_t image[][kDisplayWidth + kSideBarWidth][3],
+                                     uint8_t occupancyMask[][kDisplayWidth + kSideBarWidth], bool drawUndefinedArea) {
+	if (!image) {
+		return true;
+	}
+
+	if (!occupancyMask) {
+		return true;
+	}
+
+	PadLEDs::renderingLock = true;
+
+	// erase current image as it will be refreshed
+	memset(image, 0, sizeof(uint8_t) * kDisplayHeight * (kDisplayWidth + kSideBarWidth) * 3);
+
+	// erase current occupancy mask as it will be refreshed
+	memset(occupancyMask, 0, sizeof(uint8_t) * kDisplayHeight * (kDisplayWidth + kSideBarWidth));
+
+	//render midi view
+	for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
+
+		uint8_t* occupancyMaskOfRow = occupancyMask[yDisplay];
+		int32_t imageWidth = kDisplayWidth + kSideBarWidth;
+
+		renderRow(&image[0][0][0] + (yDisplay * imageWidth * 3), occupancyMaskOfRow, yDisplay);
+	}
+
+	PadLEDs::renderingLock = false;
+
+	return true;
+}
+
+/// render every column, one row at a time
+void MidiSessionView::renderRow(uint8_t* image, uint8_t occupancyMask[], int32_t yDisplay) {
```

Comment:
explain that this is drawing the shortcut pads on the grid and their learned status